presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -2840,15 +2840,28 @@ public ConnectorTableLayout getTableLayout(ConnectorSession session, ConnectorTa
         }
 
         TupleDomain<ColumnHandle> predicate;
+        RowExpression subfieldPredicate;
         if (hiveLayoutHandle.isPushdownFilterEnabled()) {
             predicate = hiveLayoutHandle.getDomainPredicate()
                     .transform(subfield -> isEntireColumn(subfield) ? subfield.getRootName() : null)
                     .transform(hiveLayoutHandle.getPredicateColumns()::get)
                     .transform(ColumnHandle.class::cast)
                     .intersect(createPredicate(partitionColumns, partitions));
+
+            // capture subfields from domainPredicate to add to remainingPredicate
+            // so those filters don't get lost
+            Map<String, Type> columnTypes = hiveColumnHandles(table).stream()
+                    .collect(toImmutableMap(HiveColumnHandle::getName, columnHandle -> columnHandle.getColumnMetadata(typeManager).getType()));
+            SubfieldExtractor subfieldExtractor = new SubfieldExtractor(functionResolution, rowExpressionService.getExpressionOptimizer(), session);
+
+            subfieldPredicate = rowExpressionService.getDomainTranslator().toPredicate(
+                    hiveLayoutHandle.getDomainPredicate()
+                            .transform(subfield -> !isEntireColumn(subfield) ? subfield : null)
+                            .transform(subfield -> subfieldExtractor.toRowExpression(subfield, columnTypes.get(subfield.getRootName()))));
```

Comment:
I thought there was a separate flag for subfield pushown. Do we need to check that here? Also, if it happens after the planning/optimization these info should be already available here?

Reply:
There are two different optimizations
1. pushing down filters - this includes filters on subfields
2. subfield pushdown  - this means telling the reader to prune out all the fields from the struct that don't get used by the query

the relevant optimization here is filter pushdown.  I'm actually not sure why we check that filter pushdown is enabled here, though. if there were no filter pushdown, there shouldn't be any filters in this field, and if something else put filters on this field we should want to know about them too.  It seems to me that filter pushdown should just create the HiveTableLayoutHandle that it wants and nobody else should need to know how it got to that state.  If that's not how it works, there's something off about the abstraction.

However, as I don't understand why the check is there in the first place, I don't feel confident removing it without more testing.  Since this is just a targeted bug fix, I kept the current logic.

Reply:
My question was more like if that's turned off will we still get this fix? Maybe just add a test by turning off subfile pushdown as well and make sure the fix still works with that.

Reply:
@kaikalur updated the test to include all permutations of these two properties.